turbopack/crates/turbo-tasks-backend/src/kv_backing_storage.rs

```diff
@@ -679,9 +679,53 @@ fn serialize_task_type(
         let deserialize: Result<CachedTaskType, _> = serde_path_to_error::deserialize(
             &mut pot_de_symbol_list().deserializer_for_slice(&*task_type_bytes)?,
         );
-        if let Err(err) = deserialize {
-            println!("Task type would not be deserializable {task_id:?}: {err:?}\n{task_type:#?}");
-            panic!("Task type would not be deserializable {task_id:?}: {err:?}");
+        match deserialize {
+            Err(err) => {
+                println!(
+                    "Task type would not be deserializable {task_id:?}: {err:?}\n{task_type:#?}"
+                );
+                panic!("Task type would not be deserializable {task_id:?}: {err:?}");
+            }
+            Ok(task_type2) => {
+                if &task_type2 != task_type {
+                    println!(
+                        "Task type would not round-trip {task_id:?}:\noriginal: \
+                         {task_type:#?}\nround-tripped: {task_type2:#?}"
+                    );
+                    panic!(
+                        "Task type would not round-trip {task_id:?}:\noriginal: \
+                         {task_type:#?}\nround-tripped: {task_type2:#?}"
+                    );
+                }
+                let mut bytes2 = Vec::new();
+                let result2 = POT_CONFIG.serialize_into(&task_type2, &mut bytes2);
+                match result2 {
+                    Err(err) => {
+                        println!(
+                            "Task type would not be serializable the second time {task_id:?}: \
+                             {err:?}\n{task_type2:#?}"
+                        );
+                        panic!(
+                            "Task type would not be serializable the second time {task_id:?}: \
+                             {err:?}\n{task_type2:#?}"
+                        );
+                    }
+                    Ok(()) => {
+                        if bytes2 != *task_type_bytes {
```

Comment:
This is assuming a value will serialize the same way every time? What if there's a `HashMap`? It's valid for that `HashMap` to use a random hash seed every time it's deserialized. That's okay as long as we don't depend on iteration order.

I think `FxHashMap` is using a zero-seeded map, but we might use a normal `HashMap` in a few places.

Reply:
Yep. It should check that it serializes to the same bytes. This is required as we use the serialized bytes to lookup the entry in the fs cache.

HashMaps would break this invariant and should fail this check. They also don't implement TaskInput so they can't be used as task key.